### ISSUE TITLE
Splitter: Check keyword (typos) when a splitter is created.

### ISSUE THREAD

**Issue by dafrose** [nipype/pydra#531]
# What would you like changed/added and why?

I would like `pydra` to check, if keywords in a splitter exist/are typed correctly, on the line where the splitter is set in contrast to when it is actually implemented during runtime. 

# What would be the benefit? Does the change make something easier to use?

1. A typo can happen very easily and is very simple to correct, but difficult to identify if the error message does not point to the place, where it actually happens. This is especially true for users with less programming experience. Since the error is only caught, when `pydra` is trying to apply the splitter, not when it is created, the error traceback does not point to the place, where it happens. This makes debugging more difficult, especially in large, nested workflows, where it already requires some work and experience to find the original error message. 
2. In the current implementation, a long-running workflow with many splitters and combiners can run for a while, before the mistake is noticed and caught. To increase development efficiency, it would be great to catch the mistake before any tasks are run.

# Examples
I did this in a jupyter notebook, so `nest_asyncio` might not be necessary if run as a script.
```python
# preamble
import nest_asyncio
nest_asyncio.apply()
import pydra

@pydra.mark.task
@pydra.mark.annotate({"return": {"a_plus_two": float,
                           }})
def add_two(a: float = 0):
    a_plus_two = a + 2
    print(f"{a} + 2 = {a_plus_two}")
    return a_plus_two

@pydra.mark.task
@pydra.mark.annotate({"return": {"a_plus_b": float,
                           }})
def add_any(a: float = 0, b: float = 0):
    a_plus_b = a + b
    print(f"{a} + {b} = {a_plus_b}")
    return a_plus_b

numbers = [1,2,3,4,5]
```

## Example 1: Simple Function Task

```python
my_adder = add_two(a=numbers)
my_adder.split("b")
my_adder()
my_adder.result()
```
Result:

```python
KeyError                                  Traceback (most recent call last)
/tmp/ipykernel_3684442/2629494607.py in <module>
----> 1 my_adder()
      2 my_adder.result()

# [... skipping 9 entries]

/.../lib/python3.7/site-packages/pydra/engine/state.py in _single_op_splits(self, op_single)
   1072         """splits function if splitter is a singleton"""
   1073         shape = hlpst.input_shape(
-> 1074             self.inputs[op_single], cont_dim=self.cont_dim.get(op_single, 1)
   1075         )
   1076         val_ind = range(reduce(lambda x, y: x * y, shape))

KeyError: 'add_two.b'
```

While this error message is very long, it can be understood fairly easy in this case, because there is not much to look for and the `KeyError` already contains the most important info.

## Example 2: Workflow with two splitters

In this example, the mistake happens in the second function within a workflow. The original workflow already contains a splitter as does the second function.

```python 
workflow = pydra.Workflow(name="adder_workflow",
                          input_spec=["a", "b"],
                          a=numbers,
                          b=numbers
                          )
workflow.split("a")
workflow.add(add_two(name="my_add_two",
                     a=workflow.lzin.a))

workflow.add(add_any(name="my_add_any",
                     a=workflow.my_add_two.lzout.a_plus_two,
                     b=workflow.lzin.b
                    ))
workflow.my_add_any.split("b_")  # this is where the mistake happens

workflow.set_output([("my_output", workflow.my_add_any.lzout.a_plus_b),
                    ])

with pydra.Submitter(plugin="cf") as sub:
    sub(workflow)
```

Result:

```python
3 + 2 = 5
1 + 2 = 32 + 2 = 4
4 + 2 = 6
5 + 2 = 7

---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
/tmp/ipykernel_3684442/199430367.py in <module>
     18 
     19 with pydra.Submitter(plugin="cf") as sub:
---> 20     sub(workflow)

# [... skipping 16 (!) entries]

/.../lib/python3.7/site-packages/pydra/engine/state.py in _single_op_splits(self, op_single)
   1072         """splits function if splitter is a singleton"""
   1073         shape = hlpst.input_shape(
-> 1074             self.inputs[op_single], cont_dim=self.cont_dim.get(op_single, 1)
   1075         )
   1076         val_ind = range(reduce(lambda x, y: x * y, shape))

KeyError: 'my_add_any.b_'
```

Now the problem here is, that (a) other parts of the workflow already ran, potentially for many hours, before it stopped, possibly when nobody is there to notice it. (b) The error message is considerably longer and points at a point that does not give any information whatsoever at where the actual mistake happened. The error message might seem quite clear, but in a more complex situation, it might actually be quite ambiguous as to what it refers to (e.g. similar names at different points in the workflow, etc.). Also, in nested workflows in a script (not interactive shell) the message may be hidden behind many more layers of errors. 
​

# Conclusion

I think it would be best to check the keyword exactly, when `.split()` is called. I'll try to find the time to come up with a code proposal, but would appreciate your feedback first.

**Comment by dafrose** [comment]
@djarecka I think there is something about that example, that I do not understand. Available variable names/attributes of a function task are derived via introspection of the original function, do they not? Or in case of another type of task, via the previously defined `input_spec`. So does/should it matter whether the value of the name is set before/after the call to `.split()`. From what I understand about how the inputs are parsed, any names created via inputs are compared to the list of actual function arguments and anything that does not exist in the function declaration is simply dropped (which is the basis for issue #262 ). Thus, it shouldn't matter when a particular value is assigned via `.inputs.<name>`, even if that creates a new attribute, because any name used in `.split()` only ever needs to be compared against the function declaration/`input_spec` of a task. Am I wrong?

**Comment by djarecka** [comment]
You're right. I guess we should be checking earlier. Thank you for pointing this.
